rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,288 @@
+    state.motionState === 'exiting' && styles.exiting,
+    styles.root,
+  );
+
+  return state;
+};
+```
+
+Styles, in case of CSS animations:
+
+```ts
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import type { SampleSlots, SampleState } from './Sample.types';
+import type { SlotClassNames } from '@fluentui/react-utilities';
+import { tokens } from '@fluentui/react-theme';
+
+export const SampleClassNames: SlotClassNames<SampleSlots> = {
+  root: 'fui-Sample',
+};
+
+/**
+ * Styles for the root slot
+ */
+const visibleKeyframe = {
+  ...shorthands.borderRadius(0),
+  opacity: 1,
+  transform: 'translate3D(0, 0, 0) scale(1)',
+};
+
+const hiddenKeyframe = {
+  ...shorthands.borderRadius('36px'),
+  opacity: 0,
+  transform: 'translate3D(-100%, 0, 0) scale(0.9)',
+};
+
+const useStyles = makeStyles({
+  root: {
+    willChange: 'opacity, transform, border-radius',
+  },
+
+  entering: {
+    animationDuration: '500ms',
+    animationTimingFunction: tokens.curveDecelerateMid,
+    animationName: {
+      '0%': hiddenKeyframe,
+      '100%': visibleKeyframe,
+    },
+  },
+
+  exiting: {
+    animationDuration: '750ms',
+    animationTimingFunction: tokens.curveAccelerateMin,
+    animationName: {
+      '0%': visibleKeyframe,
+      '100%': hiddenKeyframe,
+    },
+  },
+});
+
+export const useSampleStyles_unstable = (state: SampleState): SampleState => {
+  const styles = useStyles();
+
+  state.root.className = mergeClasses(
+    SampleClassNames.root,
+    state.motionState === 'entering' && styles.entering,
+    state.motionState === 'exiting' && styles.exiting,
+    styles.root,
+  );
+
+  return state;
+};
+```
+
+Overriding the transition/animation on the Application side:
+
+```tsx
+import * as React from 'react';
+import { Drawer } from '@fluentui/react-drawer';
+import { makeStyles } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  customDuration: {
+    transitionDuration: '500ms',
+    animationDuration: '750ms',
+  },
+});
+
+export const CustomDuration = () => {
+  const styles = useStyles();
+
+  return <Drawer className={styles.customDuration} />;
+};
+```
+
+#### How it works?
+
+The hook accepts a boolean value that represents the "presence" state. It internally tracks the animation/transition and return boolean values related to its motion state and a ref that needs to be assigned to the element we are tracking.
+
+```ts
+const {
+  /**
+   * Ref used to track the target element that requires animation. It should
+   * be passed to the element responsible for performing the animation.
+   */
+  ref,
+
+  /**
+   * Determines whether the component should be displayed on the screen.
+   *
+   * This property will evaluate to `true` when the provided presence value is true.
+   * Once the provided value changes to false, the hook will monitor the transition/animation
+   * completion and subsequently set this property to false.
+   */
+  shouldRender,
```

Comment:
I wonder if `shouldRender` and `visible` should be collapsed into a single value like we did for `motionState`? Something like: `visible: boolean`?

I'm finding it hard to think about the difference between `visible` and `shouldRender`. I get the distinction after reading the source and understand why it's useful to track both internally but are both props useful to a user of the hook?

Looking at your `Drawer` PR I think `shouldRender` is always derived from the `presence` value that is passed in to `useMotionPresence()` so it should be something a consumer of this hook already knows.

What do you think?

Reply:
A detailed explanation:

Using the example of an element that can appear and disappear from the screen (like a Dialog or Drawer) and uses CSS Transitions (not animations). 
When React freshly create the DOM element (on mount), the CSS transition can't kick in, because the element is already fully rendered based on the last state. To be able to apply CSS transition, we then need to apply a different class that will contain styles to represent the last UI "state" of that transition. 

For that case, we then need two booleans. One to say "we can render/remove the DOM node" and the other "we can apply/remove the CSS class to add transition". That's the difference between `shouldRender` and `visible`.

`shouldRender` is a boolean that represents when an element can be rendered but also when it can be safely removed. It is `true` when `presence` becomes `true` but will only be `false` after all animations/transitions finish, so we know it's safe to unmount.

`visible` on the other hand only turns `true` on the very next frame after the element is rendered, so we can kick in the CSS transitions by applying a different class. It'll become `false` when presence is also false, so the transition can go into reverse. When the transition ends, then `shouldRender` will be false.

With that, `shouldRender` and `visible` returned values are different from `presence` parameter and therefore cannot be removed. If we collapse them into one the code on the render layer will be a little much more complicated:
```tsx
  if (!state.shouldRender) {
    return null;
  }

  return <slots.root {...slotProps.root} />;
```
vs
```tsx
  if (!state.renderState === 'shouldRender' || !state.renderState === 'visible') {
    return null;
  }

  return <slots.root {...slotProps.root} />;
```

All of this is for Transitions. Animations won't ever need the `visible` value, as they can start when the element is first rendered. 

A full comparison between how to apply Transition/Animation:

CSS Transition
```ts
const useStyles = makeStyles({
  root: {
    opacity: 0,
    transitionProperty: 'opacity',
    transitionDuration: '200ms',
  },

  visible: {
    opacity: 1,
  },
});

export const useSampleStyles_unstable = (state: SampleState): SampleState => {
  const styles = useStyles();

  state.root.className = mergeClasses(
    /* 
     * When the "shouldRender" is true, the element will have `opacity: 0`
     */
    styles.root,
    
    /* 
     * For the first render, when the next frame happens, a rerender occurs 
     * applying `opacity: 1` and the transition will start
     */
    state.visible && styles.visible,      	
  );

  return state;
};
```

Animation:
```js
const useStyles = makeStyles({
  root: {
  	opacity: 0,
    animationDuration: '500ms',
  },

  entering: {
    animationName: {
	  to: {
		opacity: 0
	  }
    },
  },

  exiting: {
    animationName: {
	  to: {
		opacity: 1
	  }
    },
  },
});

export const useSampleStyles_unstable = (state: SampleState): SampleState => {
  const styles = useStyles();

  state.root.className = mergeClasses(
    SampleClassNames.root,
    // We don't use `visible` here. We only differentiate the "entering" from "exiting" UI states
    state.motionState === 'entering' && styles.entering,
    state.motionState === 'exiting' && styles.exiting,
    styles.root,
  );

  return state;
};
```

Reply:
Thanks for the explanation, it really helps!

> shouldRender is a boolean that represents when an element can be rendered but also when it can be safely removed. It is true when presence becomes true but will only be false after all animations/transitions finish, so we know it's safe to unmount.

Is it correct that on my transition in (i.e., when `Drawer` mounts) `shouldRender` becomes `false` after the transition ends? Even though we should continue to render `Drawer` since it's opening/appearing in the UI?

If that's the case perhaps "shouldRender" isn't the best name here since the value doesn't really represent whether something should render so much as if it's currently in a transitory state? Maybe something like `isAnimating/Transitioning`?

Assuming my understanding above is correct, this next issue can be ignored but I'll add it just for completeness :)

My concerns with `shouldRender` and `visible` are:

1. Without diving into the code or getting a more detailed explanation they feel like the same thing. If it were just perusing the API docs I think I would be confused by them.
2. With the existing names it seems I can express contradictory states like `shouldRender = false` when `visible = true`. If `shouldRender` is really something like `isTransitioning` this concern goes away.
3. It feels like we're putting more work on the users of this hook than we might need to (your explanation has me leaning toward "we need to") in having to manage both `shouldRender` and `visible`. I implemented a similar API for [`SpinButton` that requires users to control both `value` and `displayValue` to get custom formatting](https://react.fluentui.dev/?path=/docs/components-spinbutton--default#display-value). Not exactly the same situation as here but I've never really been happy with this `SpinButton` API.

---

I don't really have a strong objection to having both `shouldRender` and `visible`, other than that we may want to rename `shouldRender`, I'm more just pushing for us to spend some time considering if we can simplify the API further. If not, that's fine by me :)

Reply:
-- shouldRender is a boolean that represents when an element can be rendered but also when it can be safely removed. It is true when presence becomes true but will only be false after all animations/transitions finish, so we know it's safe to unmount. 

What I meant here was that `shouldRender` become false when all **exit** animations finish ( triggered by `presence` false). `visible` will never be true if `shouldRender` is false. 

We can consider renaming the visible flag to something else, like `appear`. 

What do you think?

Reply:
I made some changes to `shouldRender` and `visible` flags.
- `active` got renamed to `active`, which means the element is rendered and "active" on the screen.
- `shouldRender` got removed as we can use `motionState === 'unmounted'` instead.

This might make it clearer the intent behind each prop.